pyccel/ast/builtins.py

```diff
@@ -262,13 +262,25 @@ def __new__(cls, args):
     def dtype(self):
         return 'tuple'
 
+    @property
+    def homogeneous_dtype(self):
+        assert(self.is_homogeneous)
+        return self._homogeneous_dtype
+
     @property
     def shape(self):
-        return [len(self._args)]
+        if (self._arg_dtypes is None):
+            return [len(self._args)]
+        else:
+            shape = [len(self._args)]
+            if self.is_homogeneous and isinstance(self._arg_dtypes[0]['datatype'], NativeTuple):
+                shape = shape + list(self._args[0].shape)
+
+            return tuple(shape)
 
     @property
     def rank(self):
-        return 1 if self.is_homogeneous else 0
+        return max(a.rank if hasattr(a, 'rank') else 0 for a in self._args)+1
```

Comment:
Here `a.rank if hasattr(a, 'rank') else 0` may be written as `getattr(a, 'rank', 0)`, which is more concise.